File: _inc/laravel/app/Http/Controllers/Configs/PipelineController.php
<?php

namespace App\Http\Controllers\Configs;

use App\Http\Controllers\Abstracts\Controller;

use App\Config\Constants\{
  ActivitiesConstants as AC,
  DatabaseConstants as DC,
  PermissionsConstants as PMC,
  ProjectsConstants as PJC,
  MiddlewaresConstants as MWC,
  UsersConstants as UC,
  ViewsConstants as VW
};
use App\Models\{
  ActivityLog,
  ClientDeal,
  Deal,
  DealDiscussion,
  DealFile,
  DealTask,
  Pipeline,
  UserDeal
};
use App\Traits\{ChecksLogin, ChecksPermissions};
use Illuminate\Http\{
  JsonResponse,
  RedirectResponse,
  Request
};
use Illuminate\Support\Facades\{
  Auth,
  Log,
  Validator,
  View as ViewFacade,
};
use Illuminate\Contracts\Support\Renderable;

use function App\Http\Controllers\Helpers\defaultPermissionDenial;
use App\Traits\DefinesResourceActions;
final class PipelineController extends Controller
{
	use DefinesResourceActions;

  use ChecksLogin, ChecksPermissions;

  private const REDIRECT_INDEX = '/';
  public const IDX = 'index';
  public const CRT = 'create';
  public const STR = 'store';
  public const SHW = 'show';
  public const EDT = 'edit';
  public const UPD = 'update';
  public const DEL = 'destroy';

  public function __construct()
  {
    $this->middleware([MWC::AUTH, MWC::XSS]);
  }

  public function index(Request $req): Renderable|RedirectResponse
  {
    $action = class_basename(static::class) . '@' . __FUNCTION__;
    $function = __FUNCTION__;
    return $this->measureProfile($action, function () use ($req, $action, $function) {
      if (($u = self::_checkLogin()) instanceof RedirectResponse) return $u;
      if (($c = self::guard($req, PMC::MNG_PPL, self::REDIRECT_INDEX)) !== true) return $c;
      $pipelines = Pipeline::where(DC::COL_TABLE_CREATOR, $u->creatorId())->get();
      Log::info($action . ' fetched', ['count' => $pipelines->count(), UC::COL_USER_ID => $u->id]);
      return ViewFacade::make(VW::PPL . '.' . $function, compact('pipelines'));
    }, ['uri' => $req->getRequestUri(), 'ip' => $req->ip()]);
  }

  public function create(Request $req): Renderable|RedirectResponse
  {
    $action = class_basename(static::class) . '@' . __FUNCTION__;
    $function = __FUNCTION__;
    return $this->measureProfile($action, function () use ($req, $function) {
      if (($u = self::_checkLogin()) instanceof RedirectResponse) return $u;
      if (($c = self::guard($req, 'create pipeline', self::REDIRECT_INDEX)) !== true) return $c;
      return ViewFacade::make(VW::PPL . '.' . $function);
    }, ['uri' => $req->getRequestUri(), 'ip' => $req->ip()]);
  }

  public function show(Request $request, Pipeline $pipeline): RedirectResponse
  {
    $action = class_basename(static::class) . '@' . __FUNCTION__;
    return $this->measureProfile($action, function () use ($request, $pipeline, $action) {
      Log::info("$action start", [PJC::COL_PPL_ID => $pipeline->id, UC::COL_USER_ID => Auth::id()]);
      if (($user = self::_checkLogin()) instanceof RedirectResponse) return $user;
      if (($c = self::guard($request, PMC::MNG_PPL, self::REDIRECT_INDEX)) !== true) {
        Log::warning("$action permission denied", [UC::COL_USER_ID => $user?->id]);
        return $c;
      }
      if ($pipeline[DC::COL_TABLE_CREATOR] !== $user?->creatorId()) {
        Log::warning("$action ownership denied", [UC::COL_USER_ID => $user?->id, PJC::COL_PPL_ID => $pipeline->id]);
        return defaultPermissionDenial($request, null, $action);
      }
      Log::info("$action redirecting", [PJC::COL_PPL_ID => $pipeline->id]);
      return redirect()->route(VW::PPL . '.index');
    }, ['uri' => $request->getRequestUri(), 'ip' => $request->ip()]);
  }

  public function store(Request $req): RedirectResponse|JsonResponse
  {
    $action = class_basename(static::class) . '@' . __FUNCTION__;
    return $this->measureProfile($action, function () use ($req) {
      if (($u = self::_checkLogin()) instanceof RedirectResponse) return $u;
      if (($c = self::guard($req, 'create pipeline', self::REDIRECT_INDEX)) !== true) return $c;
      if ($c = self::v($req, [PJC::COL_PPL_NM => 'required|max:20'])) return $c;
      Pipeline::create([
        PJC::COL_PPL_NM => $req->name,
        DC::COL_TABLE_CREATOR => $u->creatorId()
      ]);
      return redirect()->route(VW::PPL . '.index')->with('success', __('Pipeline successfully created!'));
    }, ['uri' => $req->getRequestUri(), 'ip' => $req->ip()]);
  }

  public function edit(Request $req, Pipeline $pipeline): Renderable|RedirectResponse
  {
    $action = class_basename(static::class) . '@' . __FUNCTION__;
    $function = __FUNCTION__;
    return $this->measureProfile($action, function () use ($req, $pipeline, $function) {
      if (($u = self::_checkLogin()) instanceof RedirectResponse) return $u;
      if (($c = self::guard($req, 'edit pipeline', self::REDIRECT_INDEX)) !== true) return $c;
      if ($pipeline[DC::COL_TABLE_CREATOR] !== $u->creatorId()) return defaultPermissionDenial($req, new \Exception('owner'));
      return ViewFacade::make(VW::PPL . '.' . $function, compact('pipeline'));
    }, ['uri' => $req->getRequestUri(), 'ip' => $req->ip()]);
  }

  public function update(Request $req, Pipeline $pipeline): RedirectResponse|JsonResponse
  {
    $action = class_basename(static::class) . '@' . __FUNCTION__;
    return $this->measureProfile($action, function () use ($req, $pipeline) {
      if (($u = self::_checkLogin()) instanceof RedirectResponse) return $u;
      if (($c = self::guard($req, 'edit pipeline', self::REDIRECT_INDEX)) !== true) return $c;
      if ($pipeline[DC::COL_TABLE_CREATOR] !== $u->creatorId()) return redirect()->back()->with('error', __('Permission Denied.'));
      if ($c = self::v($req, [PJC::COL_PPL_NM => 'required|max:20'])) return $c;
      $pipeline->update([PJC::COL_PPL_NM => $req->name]);
      return redirect()->route(VW::PPL . '.index')->with('success', __('Pipeline successfully updated!'));
    }, ['uri' => $req->getRequestUri(), 'ip' => $req->ip()]);
  }

  public function destroy(Request $req, Pipeline $pipeline): RedirectResponse|JsonResponse
  {
    $action = class_basename(static::class) . '@' . __FUNCTION__;
    return $this->measureProfile($action, function () use ($req, $pipeline) {
      if (($u = self::_checkLogin()) instanceof RedirectResponse) return $u;
      if (($c = self::guard($req, 'delete pipeline', self::REDIRECT_INDEX)) !== true) return $c;
      if ($pipeline[DC::COL_TABLE_CREATOR] !== $u->creatorId()) return redirect()->back()->with('error', __('Permission Denied.'));
      if ($pipeline->stages->isNotEmpty()) return redirect()->route(VW::PPL . '.index')->with('error', __('There are stages or deals in this pipeline, remove them first!'));
      foreach ($pipeline->stages as $stage)
        foreach (Deal::where([[PJC::COL_PPL_ID, $pipeline->id], ['stage_id', $stage->id]])->get() as $deal) {
          DealDiscussion::where(AC::COL_DL, $deal->id)->delete();
          DealFile::where(AC::COL_DL, $deal->id)->delete();
          ClientDeal::where(AC::COL_DL, $deal->id)->delete();
          UserDeal::where(AC::COL_DL, $deal->id)->delete();
          DealTask::where(AC::COL_DL, $deal->id)->delete();
          ActivityLog::where(AC::COL_DL, $deal->id)->delete();
          $deal->delete();
        }
      $pipeline->delete();
      return redirect()->route(VW::PPL . '.index')->with('success', __('Pipeline successfully deleted!'));
    }, ['uri' => $req->getRequestUri(), 'ip' => $req->ip()]);
  }

  private static function v(Request $r, array $rules): ?RedirectResponse
  {
    $v = Validator::make($r->all(), $rules);
    return $v->fails() ? redirect()->back()->with('error', $v->getMessageBag()->first()) : null;
  }
}

File: _inc/_DEPRECATED_django/erp_prestech_backend/app/Http/Controllers/configs/pipeline_controller.py
import inspect
import logging
from django.contrib import messages
from django.core.exceptions import PermissionDenied
from django.http import HttpRequest, HttpResponse
from django.shortcuts import render, redirect, get_object_or_404
from .._helpers.error_handlers import default_permission_denial, default_undefined_exception
from .._helpers.http import get_redirect_url
from .._traits.controller import Controller
from ....Models.configs.pipeline import Pipeline
from ....Models.activity.deal import Deal
from ....Models.activity.deal_discussion import DealDiscussion
from ....Models.activity.deal_file import DealFile
from ....Models.activity.client_deal import ClientDeal
from ....Models.activity.user_deal import UserDeal
from ....Models.planning.deal_task import DealTask
from ....Models.activity.activity_log import ActivityLog

logger = logging.getLogger(__name__)

class PipelineController(Controller):
  @classmethod
  def index(cls, request: HttpRequest) -> HttpResponse:
    REF=f'{cls.__name__}::{inspect.currentframe().f_code.co_name}'
    self = cls() 
    self.request = request
    try:
      self.authorize('manage_pipeline')
      pipelines = Pipeline.objects.filter(created_by=request.user.creator_id())
      return render(request, 'pipelines/index.html', {'pipelines': pipelines})
    except PermissionDenied as e:
      return default_permission_denial(
        request,
        err=e,
        ref=REF,
        logger=logger
      )
    except Exception as e:
      return default_undefined_exception(
        request,
        err=e,
        ref=REF,
        logger=logger
      )

  @classmethod
  def create(cls, request: HttpRequest) -> HttpResponse:
    REF=f'{cls.__name__}::{inspect.currentframe().f_code.co_name}'
    self = cls() 
    self.request = request
    try:
      self.authorize('create_pipeline')
      return render(request, 'pipelines/create.html')
    except PermissionDenied as e:
      return default_permission_denial(
        request,
        err=e,
        ref=REF,
        logger=logger,
        json={'error': 'PermissionDenied.'}
      )
    except Exception as e:
      return default_undefined_exception(
        request,
        err=e,
        ref=REF,
        logger=logger
      )

  @classmethod
  def store(cls, request: HttpRequest) -> HttpResponse:
    REF=f'{cls.__name__}::{inspect.currentframe().f_code.co_name}'
    self = cls() 
    self.request = request
    try:
      self.authorize('create_pipeline')
      name = request.POST.get('name', '').strip()
      if not name or len(name) > 20:
        messages.error(request, 'Name is required and must be <= 20 characters.')
        return redirect('pipelines.index')
      pipeline = Pipeline(name=name, created_by=request.user.creator_id()) 
      pipeline.save()
      messages.success(request, 'Pipeline successfully created!')
      return redirect('pipelines.index')
    except PermissionDenied as e:
      return default_permission_denial(
        request,
        err=e,
        ref=REF,
        logger=logger
      )
    except Exception as e:
      return default_undefined_exception(
        request,
        err=e,
        ref=REF,
        logger=logger
      )

  @classmethod
  def show(cls, request: HttpRequest, pipeline_id: str) -> HttpResponse:
    return redirect('pipelines.index')

  @classmethod
  def edit(cls, request: HttpRequest, pipeline_id: str) -> HttpResponse:
    REF=f'{cls.__name__}::{inspect.currentframe().f_code.co_name}'
    self = cls() 
    self.request = request
    try:
      self.authorize('edit_pipeline')
      pipeline = get_object_or_404(Pipeline, pk=pipeline_id)
      if pipeline.created_by != request.user.creator_id():
        return default_permission_denial(
							request,
							err='ID Blocked.',
							ref=REF,
							logger=logger,
							json={'error': 'PermissionDenied.'}
						)
      return render(request, 'pipelines/edit.html', {'pipeline': pipeline})
    except PermissionDenied as e:
      return default_permission_denial(
        request,
        err=e,
        ref=REF,
        logger=logger,
        json={'error': 'PermissionDenied.'}
      )
    except Exception as e:
      return default_undefined_exception(
        request,
        err=e,
        ref=REF,
        logger=logger
      )

  @classmethod
  def update(cls, request: HttpRequest, pipeline_id: str) -> HttpResponse:
    REF=f'{cls.__name__}::{inspect.currentframe().f_code.co_name}'
    self = cls() 
    self.request = request
    try:
      self.authorize('edit_pipeline')
      pipeline = get_object_or_404(Pipeline, pk=pipeline_id)
      if pipeline.created_by != request.user.creator_id():
        messages.error(request, 'Permission denied.')
        return redirect(get_redirect_url(request))
      name = request.POST.get('name', '').strip()
      if not name or len(name) > 20:
        messages.error(request, 'Name is required and must be <= 20 characters.')
        return redirect('pipelines.index')
      pipeline.name = name 
      pipeline.save()
      messages.success(request, 'Pipeline successfully updated!')
      return redirect('pipelines.index')
    except PermissionDenied as e:
      return default_permission_denial(
        request,
        err=e,
        ref=REF,
        logger=logger
      )
    except Exception as e:
      return default_undefined_exception(
        request,
        err=e,
        ref=REF,
        logger=logger
      )

  @classmethod
  def destroy(cls, request: HttpRequest, pipeline_id: str) -> HttpResponse:
    REF = f'{cls.__name__}::{inspect.currentframe().f_code.co_name}'
    self = cls(); self.request = request
    try:
      self.authorize('delete_pipeline');
      pipeline = get_object_or_404(Pipeline, pk=pipeline_id)
      if pipeline.created_by != request.user.creator_id():
        messages.error(request, 'Permission denied.');
        return redirect(get_redirect_url(request))
      if pipeline.stages.count() == 0:
        for stage in pipeline.stages.all():
          deals = Deal.objects.filter(pipeline_id=pipeline.id, stage_id=stage.id)
          for deal in deals:
            for _cls in (DealDiscussion, DealFile, ClientDeal,
                         UserDeal, DealTask, ActivityLog):
              _cls.objects.filter(deal_id=deal.id).delete()
            deal.delete()
          stage.delete()
        pipeline.delete()
        messages.success(request, 'Pipeline successfully deleted!');
      else:
        messages.error(
          request,
          'There are some Stages and Deals on Pipeline, please remove it first!'
        );
      return redirect('pipelines.index')
    except PermissionDenied as e:
      return default_permission_denial(request, err=e, ref=REF, logger=logger)
    except Exception as e:
      return default_undefined_exception(request, err=e, ref=REF, logger=logger)